data/completion/complete.sh

```diff
@@ -0,0 +1,73 @@
+# -*- bash -*-
+
+# _complete_from_snap serialises the tab completion request and sends it off to
+# the appropriate 'snap run --command=complete', and de-serialises the response
+# into the usual tab completion result.
+_complete_from_snap() {
+    {
+        read -a opts
+        # opts is expected to be a series of compopt options
+        if [[ ${#opts[@]} -gt 0 ]]; then
+            if [[ "${opts[0]}" == "cannot" ]]; then
+                # older snap-execs sent errors over stdout :-(
+                return 1
+            fi
+
+            for i in "${opts[@]}"; do
+                if ! [[ "$i" =~ ^[a-z]+$ ]]; then
+                    # non-alphanumeric option; something awry
+                    return 2
+                fi
+            done
+        fi
+
+        read bounced
+        case "$bounced" in
+            ""|"alias"|"export"|"job"|"variable")
+                # OK
+                ;;
+            *)
+                # unrecognised bounce
+                return 2
+                ;;
+        esac
+
+        read sep
+        if [ "$sep" ]; then
+            # non-blank separator? madness!
+            return 2
+        fi
+        local oldIFS="$IFS"
+
+        if [ ! "$bounced" ]; then
+            local IFS=$'\n'
+            COMPREPLY=( $( \grep -v '[[:cntrl:];?*{}]' ) )
```

Comment:
I meant to also say to please document the arguments to this grep. Eg:
```
# Ignore any suspicious results that are uncommon in filenames and that
# might be used to trick the user. A whitelist approach would be better
# but is impractical with UTF-8 and common characters like quotes.
```

All that said, I did the following and think we can probably do better:
```
$ mkdir foo ; cd foo
$ touch "fooz'ball" 'party "animal"' "disc (1)"
$ complexion <tab>
  |
   -> $ complexion
      disc (1)        fooz'ball       party "animal"
      $ complexion d<tab>
        |
         -> $ complexion disc (1) # tab completion has no escapes! Pressing 'enter' doesn't work as expected
            -bash: syntax error near unexpected token `('
$ ls <tab>
  |
   -> $ ls
      disc (1)        fooz'ball       party "animal"
      $ ls d<tab>
        |
         -> $ ls disc\ \(1\) # tab completion has escapes! Pressing 'enter' works as expected
            disc (1)
```

This demonstrates a usability bug. I don't think we can get away with not quoting. The trick will be not quoting the listing, but quoting the actual completion, like 'ls' does.